### PR TITLE
[ re #599 ] Overhaul of Data.Container

### PR DESCRIPTION
**Pull request by gallais**
Not touching `Data.Container.Indexed`, there's enough work as it is.

### CODE REVIEW

**Comment by gallais** [comment]
I think I'm quite happy with the current state of affairs except for this last
block in `Data.Container`:

https://github.com/agda/agda-stdlib/blob/844fb8e7e2d5c554d57a86fc91b0a1ecd81d1e55/src/Data/Container.agda#L105-L127

I don't know where this should go. :/

**Comment by gallais** [comment]
The changes are so far and wide I'm not convinced it makes sense to preserve
`Data.Container.Any` (code will be broken anyway because `Data.Container` can
re-export all the modules its code has been dispatched to because some of these
modules depend on `Data.Container` itself!).

**Comment by MatthewDaggitt** [comment]
Looks great work, thank you very much! I have a few comments but I don't know this part of the library well so forgive me if they don't make much sense:

> I don't know where this should go. :/

I think they should go in their own module somewhere in `Data.Container.Relation.Binary.Y` where I'm not quite sure what `Y` is... In `List` then `Y` = `BagAndSetEquality` but obviously that's not quite right here.

> The changes are so far and wide I'm not convinced it makes sense to preserve
`Data.Container.Any` (code will be broken anyway because `Data.Container`
can't re-export all the modules its code has been dispatched to because some
of these modules depend on `Data.Container` itself!).

Could we move the remaining contents of `Data.Container` to `Data.Container.Core`? Or is there an import cycle there?

File: src/Data/Container/Relation/Binary/Equality/Setoid.agda [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+private
+  module S = Setoid S
+open S using (_≈_) renaming (Carrier to X)
+
+------------------------------------------------------------------------
+-- Definition of equality
+
+module _ {s p} (C : Container s p) where
+
+  Eq : Rel (⟦ C ⟧ X) (e ⊔ s ⊔ p)
```
Comment (MatthewDaggitt) [inline src/Data/Container/Relation/Binary/Equality/Setoid.agda]: Is it worth it changing this to `_≋_`? Seems like a nice convention to have for equality over data that contains multiple elements (e.g. List, Vec, Container) etc.

Comment (gallais) [reply src/Data/Container/Relation/Binary/Equality/Setoid.agda]: I don't think `C` can be inferred. :/

Comment (MatthewDaggitt) [reply src/Data/Container/Relation/Binary/Equality/Setoid.agda]: Haha, my comment about not understanding about this part of the library was spot on!

**Comment by gallais** [comment]
> Could we move the remaining contents of `Data.Container` to `Data.Container.Core`? Or is there an import cycle there?

`μ` and `ν` are defined in terms of `Data.W` and `Codata.M` which import `Data.Container.Core`.
However we could add an extra layer with `Data.Container.Base`.

**Comment by MatthewDaggitt** [comment]
Thanks for all the work @gallais!

**Pull request merged**